### PR TITLE
Remove duplicate errors from an aggregate error input.

### PR DESCRIPTION
**Pull request by aveshagarwal**
This PR, in general, removes duplicate errors from an aggregate error input, and returns unique errors with their occurrence count. Specifically,  this PR helps with some scheduler errors that fill the log enormously. For example, see the following `truncated` output from a 300-plus nodes cluster, as there was a same error from almost all nodes.

[SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected., SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found.........

After this PR, the output looks like (on a 2-node cluster):

SchedulerPredicates failed due to persistentvolumeclaims "mongodb" not found, which is unexpected.(Count=2)

@derekwaynecarr @smarterclayton @kubernetes/sig-scheduling-pr-reviews 

Fixes https://github.com/kubernetes/kubernetes/issues/47145

### CODE REVIEW

**Comment by aveshagarwal** [comment]
I will add unit tests soon.

**Comment by caesarxuchao** [comment]
The code looks fine. Do we see big aggregated error other than in the scheduler?

**Comment by aveshagarwal** [comment]
I have output from only from scheduler, not from other components so can't say for sure. But it seems a high possibility. Simple grep (`grep -ri newaggregate . --exclude-dir={_output,vendor,test,kubectl} --exclude=*test.go`) shows many places where k8s.io/apimachinery/pkg/util/errors.NewAggregate is being called, though haven't verified what kind of aggregation they are.

**Comment by smarterclayton** [comment]
We use aggregate a lot - unfortunately I think it's going to have to be
case by case.  However, we *should* vet every place we send an event that
contains the error output naively and chop / limit it.

**Comment by aveshagarwal** [comment]
I have added unit tests. PTAL.

**Comment by aveshagarwal** [comment]
@k8s-bot pull-kubernetes-unit test this

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go
Comment (ncdc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: Use `fmt.Errorf` instead of `errors.New(fmt.Sprintf`

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go
Comment (ncdc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: You know that `len(errorCountMap)` is # of keys that you have, so I would do `result := make([]error, 0, len(errorCountMap))`

**Comment by aveshagarwal** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-unit test this

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go
Comment (ncdc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go]: I'd recommend using go1.7's subtests (see e.g. https://github.com/kubernetes/kubernetes/blob/0c516c3ac216cca8be8b033a712069d4d0336ec0/pkg/auth/nodeidentifier/default_test.go#L58)

**Comment by aveshagarwal** [comment]
@ncdc thanks for review. Updated. PTAL.

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go
Comment (ncdc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go]: Because you're using subtests, you don't need to include `%d` (`i`) in the output.

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go
Comment (ncdc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: I would at least put a space in between the error text and the count information. I wonder if `(Count=n)` is the best way to show this, or if something like `(repeated n times)` would look better? WDYT @timothysc @derekwaynecarr?

Also, if we keep the `Count=` format, I would make it lowercase `count`.

Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: I am fine with `(repeated n times)` too.

Comment (derekwaynecarr) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: I am fine with `(repeated n times)`  as well.

**Comment by aveshagarwal** [comment]
Following updates to this PR: 1) minor change in code to print repeated string only if count > 1. 2) changed unit tests to compare results on sorted slices of []errors, otherwise I noticed unit tests were flaking due to different orders of errors in output. 

@ncdc PTAL.

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go
Comment (ncdc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: Please add an initial space to the string

**Review by ncdc — CHANGES_REQUESTED** [review]

**Comment by aveshagarwal** [comment]
@ncdc updated, PTAL.

**Comment by ncdc** [comment]
LGTM (please squash once you get the ok from all reviewers)

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go
Comment (timothysc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: given that you already have the original err[] list from the predicates function you are doing an extra N operations here.  So it's 2N+Y where Y is the reduction.  I'd almost list to change it at the source to bin the errors.  I swear @jayunit100 did just that but it must have been in a different section of code.

**Comment by smarterclayton** [comment]
I thought that was in the scheduler for the fit error

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go
Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: If your concern is why I am calling Flatten, then the reason is that this function is not specific to scheduler and so the need to call Flatten. Also what @jayunit100 had done was for specific to the fit errors where these are other (atleast previously) unexpected errors.

Comment (timothysc) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: The point is you can bin on insertion and prevent the need to re-run through it, the append is only a couple of lines above -  

https://github.com/kubernetes/kubernetes/pull/47003/files#diff-a9420b7e5a53a445d54a716d341874abR192 

If you bin there, then you could have a NewAggregate overload that passes in the error map and you don't need a 2N+Y it's just 2Y.

Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: But as I said that this function UniqueAggregateWithCoun is not specific to scheduler.

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (timothysc) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Please bin on insertion above.  
https://github.com/kubernetes/kubernetes/pull/47003/files#diff-a9420b7e5a53a445d54a716d341874abR192

**Review by timothysc — CHANGES_REQUESTED** [review]

File: plugin/pkg/scheduler/core/generic_scheduler.go
Comment (aveshagarwal) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: I have separated functionality in 2 functions: one that calls flatten and one doesn't where it is known that the input is flattened already.

Comment (aveshagarwal) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Hold on, I am changin it another way.

**Comment by aveshagarwal** [comment]
@timothysc Updated PR. PTAL.

**Review by timothysc — APPROVED** [review]
Thx for making the change!  much better. 
/lgtm

**Review by ncdc — APPROVED** [review]

**Comment by timothysc** [comment]
I approve of the scheduler changes.. 

@caesarxuchao or @smarterclayton want to approve the api changes?

**Comment by ncdc** [comment]
@aveshagarwal can you please create a GitHub issue for this problem, then edit the description of this PR to say that it fixes said issue? Thanks!

**Comment by aveshagarwal** [comment]
@ncdc done. thanks.

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go
Comment (gyliu513) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go]: s/MessagesgCountMap/MessageCountMap

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go
Comment (gyliu513) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go]: How about s/mcp/mcm ?

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go
Comment (gyliu513) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go]: How about adding a new case only switching the order of `abc` andn `ghi` to test sort?

```
MessageCountMap{"ghi": 1, "abc": 1},
```

Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go]: As it is already merged, if you want, you could do a follow-up PR for this.

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go
Comment (aveshagarwal) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go]: sorry I saw your comment when this was already merged.

File: staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go
Comment (gyliu513) [inline staging/src/k8s.io/apimachinery/pkg/util/errors/errors_test.go]: @aveshagarwal a PR here https://github.com/kubernetes/kubernetes/pull/47206 as a follow up

**Pull request merged**
